src/dbnode/storage/series/buffer.go

```diff
@@ -181,287 +167,288 @@ func (b *dbBuffer) Write(
-	blockStart := bl.StartTime()
-	bootstrapped := false
-	for i := range b.buckets {
-		if b.buckets[i].start.Equal(blockStart) {
-			if b.buckets[i].drained {
-				return fmt.Errorf(
-					"block at %s cannot be bootstrapped by buffer because its already drained",
-					blockStart.String(),
-				)
-			}
-			b.buckets[i].bootstrap(bl)
-			bootstrapped = true
-			break
+	for iter.Next() {
+		dp, unit, annotation := iter.Current()
+		if err := encoder.Encode(dp, unit, annotation); err != nil {
+			return nil, err
 		}
 	}
-	if !bootstrapped {
-		return fmt.Errorf("block at %s not contained by buffer", blockStart.String())
+	if err := iter.Err(); err != nil {
+		return nil, err
 	}
-	return nil
-}
 
-// forEachBucketAsc iterates over the buckets in time ascending order
-// to read bucket data
-func (b *dbBuffer) forEachBucketAsc(fn func(*dbBufferBucket)) {
-	for i := 0; i < bucketsLen; i++ {
-		idx := (b.pastMostBucketIdx + i) % bucketsLen
-		fn(&b.buckets[idx])
-	}
+	return encoder.Stream(), nil
 }
 
-// computedForEachBucketAsc performs a fn on the buckets in time ascending order
-// and returns the sum of the number returned by each fn
-func (b *dbBuffer) computedForEachBucketAsc(
-	op computeBucketIdxOp,
-	fn func(now time.Time, b *dbBuffer, idx int, bucketStart time.Time) int,
-) int {
-	now := b.nowFn()
-	pastMostBucketStart := now.Truncate(b.blockSize).Add(-1 * b.blockSize)
-	bucketNum := (pastMostBucketStart.UnixNano() / int64(b.blockSize)) % bucketsLen
-	result := 0
-	for i := int64(0); i < bucketsLen; i++ {
-		idx := int((bucketNum + i) % bucketsLen)
-		curr := pastMostBucketStart.Add(time.Duration(i) * b.blockSize)
-		result += fn(now, b, idx, curr)
-	}
-	if op == computeAndResetBucketIdx {
-		b.pastMostBucketIdx = int(bucketNum)
+func (b *dbBuffer) Flush(
+	ctx context.Context,
+	blockStart time.Time,
+	id ident.ID,
+	tags ident.Tags,
+	persistFn persist.DataFn,
+	version int,
+) (FlushOutcome, error) {
+	buckets, exists := b.bucketVersionsAt(blockStart)
+	if !exists {
+		return FlushOutcomeBlockDoesNotExist, nil
 	}
-	return result
-}
 
-func (b *dbBuffer) Snapshot(ctx context.Context, blockStart time.Time) (xio.SegmentReader, error) {
-	var (
-		res xio.SegmentReader
-		err error
-	)
+	streams, err := buckets.toStreams(ctx)
+	if err != nil {
+		return FlushOutcomeErr, err
+	}
 
-	b.forEachBucketAsc(func(bucket *dbBufferBucket) {
-		if err != nil {
-			// Something already went wrong and we want to return the error to the caller
-			// as soon as possible instead of continuing to do work.
-			return
+	var stream xio.SegmentReader
+	numStreams := len(streams)
+	if numStreams == 1 {
+		stream = streams[0]
+	} else {
+		// In the majority of cases, there will only be one stream to persist
+		// here. Only when a previous flush fails midway through a shard will
+		// there be buckets for previous versions. In this case, we need to try
+		// to flush them again, so we merge them together to one stream and
+		// persist it.
```

Comment:
Hm, similarly here can we find a way to reuse the merge method and just pass options to return merged and not replace the current data there?